pygatt/bled112_backend.py

```diff
@@ -0,0 +1,1977 @@
+
+        # Log
+        self._logger.info("_ble_evt_gap_scan_response")
+        self._logger.debug("rssi = %d dBm", args['rssi'])
+        self._logger.debug("packet type = %s", packet_type)
+        self._logger.info("sender address = %s", address)
+        self._logger.debug("address type = %s", address_type)
+        self._logger.debug("data %s", str(data_dict))
+
+        # Drop locks
+        self._loglock.release()
+        self._main_thread_cond.release()
+
+    def _ble_evt_sm_bond_status(self, sender, args):
+        """
+        Handles the event for reporting a stored bond.
+
+        Adds the stored bond to the list of bond handles if no _bond_expected.
+        Sets _bonded True if _bond_expected.
+
+        sender -- Who fired the event. Should be the BGLib object. This is a
+                  product of the event system used in BGLib.
+        args -- dictionary containing the bond handle ('bond'), encryption key
+                size used in the long-term key ('keysize'), was man in the
+                middle used ('mitm'), keys stored for bonding ('keys')
+        """
+        # Get locks
+        self._main_thread_cond.acquire()
+        self._loglock.acquire()
+
+        # Add to list of stored bonds found or set flag
+        if self._bond_expected:
+            self._bond_expected = False
+            self._bonded = True
+        else:
+            self._stored_bonds.append(args['bond'])
+
+        # Notify
+        self._main_thread_cond.notify()
+
+        # Log
+        self._logger.info("_ble_evt_sm_bond_status")
+        self._logger.debug("bond handle = %s", hex(args['bond']))
+        self._logger.debug("keysize = %d", args['keysize'])
+        self._logger.debug("man in the middle = %d", args['mitm'])
+        self._logger.debug("keys = %s", hex(args['keys']))
+
+        # Drop locks
+        self._loglock.release()
+        self._main_thread_cond.release()
+
+    def _ble_evt_sm_bonding_fail(self, sender, args):
+        """
+        Handles the event for the failure to establish a bond for a connection.
+
+        Modifies _bonding_fail and _event_return. Notifies _main_thread_cond.
+
+        sender -- Who fired the event. Should be the BGLib object. This is a
+                  product of the event system used in BGLib.
+        args -- dictionary containing the return code ('result')
+        """
+        # Get locks
+        self._main_thread_cond.acquire()
+        self._loglock.acquire()
+
+        # Set flags, notify
+        self._bonding_fail = True
+        self._event_return = args['result']
+        self._main_thread_cond.notify()
+
+        # Log
+        self._logger.info("_ble_evt_sm_bonding_fail")
+        self._logger.info("Return code = %s",
+                          get_return_message(args['result']))
+
+        # Drop locks
+        self._loglock.release()
+        self._main_thread_cond.release()
+
+    # Response handlers --------------------------------------------------------
```

Comment:
These response handlers seem to share a lot of code.  Can they be abstracted into one function?